pkg/apis/storage/types.go

```diff
@@ -424,3 +445,81 @@ type CSINodeList struct {
 	// items is the list of CSINode
 	Items []CSINode
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIStorageCapacity stores the result of one CSI GetCapacity call.
+// For a given StorageClass, this describes the available capacity in a
+// particular topology segment.  This can be used when considering where to
+// instantiate new PersistentVolumes.
+//
+// For example this can express things like:
+// - StorageClass "standard" has "1234 GiB" available in "topology.kubernetes.io/zone=us-east1"
+// - StorageClass "localssd" has "10 GiB" available in "kubernetes.io/hostname=knode-abc123"
+//
+// The following three cases all imply that no capacity is available for
+// a certain combination:
+// - no object exists with suitable topology and storage class name
+// - such an object exists, but the capacity is unset
+// - such an object exists, but the capacity is zero
+//
+// The producer of these objects can decide which approach is more suitable.
+//
+// This is an alpha feature and only available when the CSIStorageCapacity feature is enabled.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning. It must be
+	// be a DNS subdomain (dots allowed, 253 characters). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>, a generated name, or a reverse-domain name which ends
+	// with the unique CSI driver name.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// NodeTopology defines which nodes have access to the storage
+	// for which capacity was reported. If not set, the storage is
+	// not accessible from any node in the cluster. If empty, the
+	// storage is accessible from all nodes.  This field is
+	// immutable.
+	//
+	// +optional
+	NodeTopology *metav1.LabelSelector
```

Comment:
What happens if a candidate node matches more than one capacity?  E.g.

```
CSIStorageCapacity x {
  StorageClassName: standard
  NodeTopology:
    MatchLabels:
      zone: east1
  Capacity: 10Gi
}

CSIStorageCapacity y {
  StorageClassName: standard
  NodeTopology:
    MatchLabels:
      region: east
  Capacity: 1Gi
}
```

In this case,one is not a superset of the other as defined.  I can maybe assume that region is a superset of zone (though that's still pending in a KEP) but storage still allows other topology keys.

what is a consumer of CSIStorageCapacity supposed to do in this case?

Reply:
That's not really a conflict. If a node has access to both of these and the volume is smaller than 10Gi, then it's fine to go ahead and pick the node. The storage system then gets to decide where exactly it provisions the volume, just as it does now for volumes with immediate binding. What matters is that the volume can be made usable by the chosen node.

A conflict would be when there are two objects with the exact same `StorageClassName` and `NodeTopology` but different capacity. This situation is not allowed and the producer is supposed to avoid it. The consumer could try to detect it, but currently doesn't because it would imply iterating over all objects instead of stopping at the first one - IMHO it's not worth the cost because it really shouldn't happen.